homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,289 @@
+
+    # Refer to the valve dev docs for device class descriptions
+    WATER = "water"
+    GAS = "gas"
+
+
+DEVICE_CLASSES_SCHEMA = vol.All(vol.Lower, vol.Coerce(ValveDeviceClass))
+
+# DEVICE_CLASS* below are deprecated as of 2021.12
+# use the ValveDeviceClass enum instead.
+DEVICE_CLASSES = [cls.value for cls in ValveDeviceClass]
+DEVICE_CLASS_WATER = ValveDeviceClass.WATER.value
+DEVICE_CLASS_GAS = ValveDeviceClass.GAS.value
+
+# mypy: disallow-any-generics
+
+
+class ValveEntityFeature(IntFlag):
+    """Supported features of the valve entity."""
+
+    OPEN = 1
+    CLOSE = 2
+    SET_POSITION = 4
+    STOP = 8
+
+
+ATTR_CURRENT_POSITION = "current_position"
+ATTR_POSITION = "position"
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Track states and offer events for valves."""
+    component = hass.data[DOMAIN] = EntityComponent[ValveEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_OPEN_VALVE, {}, "async_open_valve", [ValveEntityFeature.OPEN]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_CLOSE_VALVE, {}, "async_close_valve", [ValveEntityFeature.CLOSE]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALVE_POSITION,
+        {
+            vol.Required(ATTR_POSITION): vol.All(
+                vol.Coerce(int), vol.Range(min=0, max=100)
+            )
+        },
+        "async_set_valve_position",
+        [ValveEntityFeature.SET_POSITION],
+    )
+
+    component.async_register_entity_service(
+        SERVICE_STOP_VALVE, {}, "async_stop_valve", [ValveEntityFeature.STOP]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_TOGGLE,
+        {},
+        "async_toggle",
+        [ValveEntityFeature.OPEN | ValveEntityFeature.CLOSE],
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclass
+class ValveEntityDescription(EntityDescription):
+    """A class that describes valve entities."""
+
+    device_class: ValveDeviceClass | None = None
+
+
+class ValveEntity(Entity):
+    """Base class for valve entities."""
+
+    entity_description: ValveEntityDescription
+    _attr_current_valve_position: int | None = None
+    _attr_device_class: ValveDeviceClass | None
+    _attr_is_closed: bool | None
+    _attr_is_closing: bool | None = None
+    _attr_is_opening: bool | None = None
+    _attr_state: None = None
+    _attr_supported_features: ValveEntityFeature | None
```

Comment:
Do it like this:

https://github.com/home-assistant/core/blob/c6f9af2cabb8d1085baca23195883e5b910ac1b1/homeassistant/components/lawn_mower/__init__.py#L78